master/buildbot/process/buildrequestmerger.py

```diff
@@ -0,0 +1,248 @@
+
+from buildbot import config
+from buildbot.process.buildrequest import Priority
+from buildbot.util.lru import LRUCache
+
+
+class BuildRequestMerger(config.ReconfigurableServiceMixin, service.Service):
+    # Basic list of properties that must match for a buildrequest to be merged
+    # BuilderConfigs can define additional properties (see _propertiesMatch)
+    BASE_MERGE_PROPERTIES = ['force_rebuild', 'force_chain_rebuild']
+
+    def __init__(self, master):
+        self.master = master
+        self.properties_cache = LRUCache(
+            miss_fn=self.__propertiesCacheMissFn, max_size=20000)
+
+    def addBuildset(self,
+                    sourcestampsetid,
+                    reason,
+                    properties,
+                    triggeredbybrid=None,
+                    builderNames=None,
+                    external_idstring=None,
+                    _reactor=reactor):
+        """
+        Wrapper around buildsets.addBuildset that does merging before
+        buildrequests hit the db
+
+        ..seealso:: buildsets.addBuildset
+        """
+        start = time.time()
+
+        buildsetLog = {
+            'name': 'addBuildset',
+            'description': 'Log merges done while adding new buildsets',
+            'sourcestampsetid': sourcestampsetid,
+            'builderNames': builderNames,
+        }
+
+        codebase, branch, revision = self._getSourceStampInfo(sourcestampsetid)
+
+        buildsetLog['_getSourceStampInfo'] = {
+            'elapsed': time.time() - start,
+            'codebase': codebase,
+            'branch': branch,
+            'revision': revision,
+        }
+
+        # Find the priority for this buildset
+        priority = Priority.Default
+        if 'priority' in properties:
+            priority_property = properties.get('priority')[0]
+            priority = priority_property if priority_property and int(
+                priority_property) > 0 else Priority.Default
+
+        # For every builderName in this buildset, check which ones can be merged
+        breqsToMerge = {}
+        for builderName in builderNames:
+            builderMergeStart = time.time()
+            mergeBrid = self._getMergeBrid(
+                self.master.botmaster.builders[builderName], builderName,
+                priority, codebase, branch, revision, properties)
+            if mergeBrid:
+                breqsToMerge[builderName] = mergeBrid
+
+            buildsetLog[builderName] = {
+                'elapsed': time.time() - builderMergeStart,
+                'mergeBrid': mergeBrid
+            }
+
+        buildsetLog['elapsed'] = time.time() - start
+
+        log.msg(json.dumps(buildsetLog))
+
+        return self.master.db.buildsets.addBuildset(
+            sourcestampsetid=sourcestampsetid,
+            reason=reason,
+            properties=properties,
+            triggeredbybrid=triggeredbybrid,
+            builderNames=builderNames,
+            breqsToMerge=breqsToMerge,
```

Comment:
This one needs to be adjusted as well to get the new param self.master.db.buildsets.addBuildset